dotnet/src/Agents/Framework/Internal/BroadcastQueue.cs

```diff
@@ -0,0 +1,140 @@
+    public TimeSpan BlockDuration { get; set; } = TimeSpan.FromSeconds(1);
+
+    /// <summary>
+    /// Enqueue a set of messages for a given channel.
+    /// </summary>
+    /// <param name="channels">The target channels for which to broadcast.</param>
+    /// <param name="messages">The messages being broadcast.</param>
+    public void Enqueue(IEnumerable<ChannelReference> channels, IList<ChatMessageContent> messages)
+    {
+        lock (this._queueLock)
+        {
+            foreach (var channel in channels)
+            {
+                var queue = this.GetQueue(channel);
+                queue.Enqueue(messages);
+
+                if (!this._tasks.ContainsKey(channel.Hash))
+                {
+                    this._tasks.Add(channel.Hash, this.ReceiveAsync(channel, queue));
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Blocks until a channel-queue is not in a receive state.
+    /// </summary>
+    /// <param name="channel">A <see cref="ChannelReference"/> structure.</param>
+    /// <returns>false when channel is no longer receiving.</returns>
+    /// <throws>
+    /// When channel is out of sync.
+    /// </throws>
+    public async Task EnsureSynchronizedAsync(ChannelReference channel)
+    {
+        ChannelQueue queue;
+
+        lock (this._queueLock)
+        {
+            if (!this._queues.TryGetValue(channel.Hash, out queue))
+            {
+                return;
+            }
+        }
+
+        while (!queue.IsEmpty) // ChannelQueue is ConcurrentQueue, no need for _queueLock
+        {
+            lock (this._queueLock)
+            {
+                // Activate non-empty queue
+                if (!this._tasks.ContainsKey(channel.Hash))
+                {
+                    this._tasks.Add(channel.Hash, this.ReceiveAsync(channel, queue));
+                }
+
+                // Propagate prior failure (inform caller of synchronization issue)
+                if (this._failures.TryGetValue(channel.Hash, out var failure))
+                {
+                    this._failures.Remove(channel.Hash);
+                    throw new AgentException($"Unexpected failure broadcasting to channel: {channel.Channel.GetType().Name}", failure);
+                }
+            }
+
+            await Task.Delay(this.BlockDuration).ConfigureAwait(false);
+        }
+    }
+
+    private async Task ReceiveAsync(ChannelReference channel, ChannelQueue queue)
+    {
+        Interlocked.CompareExchange(ref this._isActive, 1, 0); // Set regardless of current state.
+
+        Exception? failure = null;
+
+        while (!queue.IsEmpty)
+        {
+            if (queue.TryPeek(out var messages)) // Leave payload on queue for retry on failure
+            {
+                try
+                {
+                    await channel.Channel.ReceiveAsync(messages).ConfigureAwait(false);
+                    queue.TryDequeue(out _); // Queue has already been peeked.  Remove head on success.
```

Comment:
If I'm reading this right then this equates to "at least once" semantics for channels receiving a given message. I'm curious if the channel itself has any de-duping in it or if there will be issues with a given message being received more than once.

Reply:
Good question.  Appreciate eyes on this code.  I'm going to first describe what's in my head and then try to answer your question.

At this point, whatever messages have been queued are taken as proper and must be recieved by the channel to remain in sync.  `Enqueue` is only exposed to the `AgentNexus` (base class for all nexus...so the usage is a closed system.  The nexus ensures that only the stream of individual messages it encounters are broadcast.  (FYI - Broadcast is a literal no-op for single channel case.)  

If the caller happens to add two of the same user message to the history...they won't be de-duped.  We don't have the visibilty to second guess caller intent.  But any single instance passed to the nexus (user or assistant/model generated) will not be duplicated due to agent processing or concurrency.

Reply:
One loose end to call-out...I'm currently working on exposing a "Flush" semantic to the `AgentChat`/`AgentNexus` that will be in next PR that takes advantage of `EnsureSynchronizedAsync`.

Reply:
Follow-up: Create sequence diagram https://github.com/microsoft/semantic-kernel/issues/5766